internal/util/flowgraph/node.go

```diff
@@ -22,62 +21,58 @@ import (
 	"github.com/milvus-io/milvus/internal/log"
 )
 
+// Node is the interface defines the behavior of flowgraph
 type Node interface {
 	Name() string
 	MaxQueueLength() int32
 	MaxParallelism() int32
 	Operate(in []Msg) []Msg
 	IsInputNode() bool
+	Start()
 	Close()
 }
 
+// BaseNode defines some common node attributes and behavior
 type BaseNode struct {
 	maxQueueLength int32
 	maxParallelism int32
 }
 
+// nodeCtx maintains the running context for a Node in flowgragh
 type nodeCtx struct {
 	node                   Node
 	inputChannels          []chan Msg
 	inputMessages          []Msg
 	downstream             []*nodeCtx
 	downstreamInputChanIdx map[string]int
 
-	NumActiveTasks    int64
-	NumCompletedTasks int64
+	closeCh chan struct{}
 }
 
-func (nodeCtx *nodeCtx) Start(ctx context.Context, wg *sync.WaitGroup) {
-	if nodeCtx.node.IsInputNode() {
-		inStream, ok := nodeCtx.node.(*InputNode)
-		if !ok {
-			log.Error("Invalid inputNode")
-		}
-		(*inStream.inStream).Start()
-	}
+// Start invoke Node `Start` method and start a worker goroutine
+func (nodeCtx *nodeCtx) Start(wg *sync.WaitGroup) {
+	nodeCtx.node.Start()
 
+	go nodeCtx.work()
+	wg.Done()
+}
+
+// work handles node work spinning
+// 1. collectMessage from upstream or just produce Msg from InputNode
+// 2. invoke node.Operate
+// 3. deliver the Operate result to downstream nodes
+func (nodeCtx *nodeCtx) work() {
 	for {
 		select {
-		case <-ctx.Done():
-			if nodeCtx.node.IsInputNode() {
-				inStream, ok := nodeCtx.node.(*InputNode)
-				if !ok {
-					log.Error("Invalid inputNode")
-				}
-				(*inStream.inStream).Close()
-				log.Debug("message stream closed",
-					zap.Any("node name", inStream.name),
-				)
-			}
-			wg.Done()
+		case <-nodeCtx.closeCh:
 			return
 		default:
 			// inputs from inputsMessages for Operate
 			inputs := make([]Msg, 0)
 
 			var res []Msg
 			if !nodeCtx.node.IsInputNode() {
-				nodeCtx.collectInputMessages(ctx)
+				nodeCtx.collectInputMessages()
```

Comment:
if channel is closed, then inputs could be a empty message list.
However in some of the nodes they check inputs length, that may cause unexpected erros

Reply:
Currently the input msg will never be empty. When closing a node, the corner case might be a list with size of input channels with nil value.